ynput/OpenPype
# Traypublisher: advanced CSV ingest

## Changelog Description
CSV data ingesting can be done through TrayPublisher in both interactive and headless mode. Settings allow for configuration of CSV columns and data expectations. Each column can be configured for data type or requirement. It is also possible to define a default value when data is missing.

This is a working prototype and doesnt include all expected features (see `Next itteration of feature plans` section of this PR ). At the moment the data in csv has to be compliled with preflight check so no validations needs to be done on them. 

_**CLI interface includes following arguments**_
- **--csv-filepath**: The absolute path to csv file which is in root of package folder
- **--project-name**: The name of a project used for ingestion
- **--asset-name**: The name of shot/asset folder used for ingestion data iteration backup
- **--task-name**: The name of task used for ingestion data iteration backup
- **--ignore-validators**: Flag argument for skipping validators

_Column configuration_
![image](https://github.com/ynput/OpenPype/assets/40640033/37089dea-b804-4f3b-ad8d-0b51d0c56320)

_Data representation configuration_
![image](https://github.com/ynput/OpenPype/assets/40640033/acd7b85b-2c98-47e9-9fba-d9dbfa1d7c22)

## Additional info
- Implementation also includes the POC prototype originally done in Standalone publisher. 
- Reopening the old [PR](https://github.com/ynput/OpenPype/pull/5740) with some cleanup and bugs the code had.

### TODO (in scope of this PR):
- [x] publishing of CSV into context of selected task (/editorial/[task] edit)
- [x] heandling representation colorspaceData (it was working in StandalonePublisher)
- [x] CLI warning message for AYON unsuported (just to avoid confusion for those who wanted to test it on AYON)
- [x] add explicit definition for head and tail. Via csv columns.
- [x] Custom Thumbnail file distribution

## Next itteration of feature plans:
- AYON addoption of the tool
- Creating missing hierarchy folders in case they are missing

## Testing notes:
1. Clone branch
2. Navigate cmd to repository folder
3. Download testing files CSV: [2023-10-2_fcg_sub0001.csv](https://github.com/ynput/OpenPype/files/12858436/2023-10-2_fcg_sub0001.csv)
4. Before you run the command make sure shots are created in your testing project (this process is not creating shot folders, it is only creating products, versions and representations)
5. adjust and use following command to run ingest. Make sure project name is the same as used in csv data.
```PowerShell
.\tools\openpype_console.bat addon traypublisher ingestcsv --csv-filepath '[CSV FILE ABS PATH]' --project-name [PROJECT_NAME] --asset-name [ASSET FOLDER NAME] --task-name [TASK NAME] --ignore-validators
```

## Dependentcy
- blocked by https://github.com/ynput/OpenPype/pull/6100
